components/fxa-client/src/fxa_client.udl

```diff
@@ -679,6 +679,24 @@ interface FirefoxAccount {
   void simulate_permanent_auth_token_issue();
 };
 
+interface FxaStateMachine {
+  // Create an FxaStateMachine
+  //
+  // Note: When restoring a connected account, only `device_config.capabilities` will be used.  We
+  // will use the device type and device name info that's stored on the server.
+  constructor(FirefoxAccount account, string oauth_entrypoint, DeviceConfig device_config);
+
+  // Get the current auth state
+  FxaState state();
+
+  // Process a auth event (login, logout, etc).
+  //
+  // On success, returns the new state.
+  // On error, the state will remain the same.
+  [Throws=FxaError]
+  FxaState process_event(FxaEvent event);
```

Comment:
What do you think about _completely_ abstracting the state machine away? i.e the state machine becomes an implementation detail, the consumer can still read the "state", because they have to for presenting UI to the user, etc but from the consumer's perspective, they are triggering procedures (or passing in events in the `process_event` case) but aren't aware of the fact there exists a state machine. Of course, you've spent a considerable time looking at the consumer code so I'd defer to your perspective there.

Reply:
This seems right to me.  My hope is that once we're done reworking the android/ios code, that consumers basically just call `process_event()`, see what new state gets returned, and can ignore all the details that happened inside the state machine.